lib/semantics/check-allocate.cc

```diff
@@ -0,0 +1,411 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "check-allocate.h"
+#include "attr.h"
+#include "expression.h"
+#include "tools.h"
+#include "type.h"
+#include "../evaluate/fold.h"
+#include "../evaluate/type.h"
+#include "../parser/parse-tree.h"
+
+namespace Fortran::semantics {
+
+struct AllocateCheckerInfo {
+  const DeclTypeSpec *typeSpec{nullptr};
+  std::optional<evaluate::DynamicType> sourceExprType;
+  bool gotStat{false};
+  bool gotMsg{false};
+  bool gotTypeSpec{false};
+  bool gotSrc{false};
+  bool gotMold{false};
+};
+
+static const parser::Name &GetName(
+    const parser::AllocateObject &allocateObject) {
+  return std::visit(
+      common::visitors{
+          [&](const parser::Name &name) -> const parser::Name & {
+            return name;
+          },
+          [&](const parser::StructureComponent &structureComponent)
+              -> const parser::Name & { return structureComponent.component; },
+      },
+      allocateObject.u);
+}
+
+class AllocationChecker {
+public:
+  AllocationChecker(
+      const parser::AllocateObject &obj, AllocateCheckerInfo &info)
+    : allocateInfo_{info}, name_{GetName(obj)}, type_{name_.symbol->GetType()},
+      isSubobject_{std::holds_alternative<parser::StructureComponent>(obj.u)} {
+    CHECK(type_ != nullptr);
+    if (type_->category() == DeclTypeSpec::Category::Character) {
+      hasDefferedTypeParameter_ =
+          type_->characterTypeSpec().length().isDeferred();
+    } else if (const DerivedTypeSpec * derivedTypeSpec{type_->AsDerived()}) {
+      for (auto it{derivedTypeSpec->parameters().begin()};
+           it != derivedTypeSpec->parameters().end(); ++it) {
+        hasDefferedTypeParameter_ |= it->second.isDeferred();
+      }
+      isAbstract_ = derivedTypeSpec->typeSymbol().attrs().test(Attr::ABSTRACT);
+    }
+    isUnlimitedPolymorphic_ =
+        type_->category() == DeclTypeSpec::Category::ClassStar;
+  }
+
+  bool RunChecks(SemanticsContext &context);
+
+private:
+  AllocateCheckerInfo &allocateInfo_;
+  const parser::Name &name_;
+  const DeclTypeSpec *type_;
+  bool isSubobject_;
+  bool hasDefferedTypeParameter_{false};
+  bool isUnlimitedPolymorphic_{false};
+  bool isAbstract_{false};
+};
+
+static std::optional<AllocateCheckerInfo> CheckAllocateOptions(
```

Comment:
Why prototype static subroutines?  If they are independent, define them here.  If they are mutually recursive, make them member functions (probably `static`) of your helper class.

Reply:
That was only a forward declaration to avoid defining them before the core functions that use them so that the reader gets to the core functions first, but it has no functional purpose and if you find this weird, I'd rather remove it. I reorganized the function order and removed these.